### PR TITLE
Resolved undocumented nullable behavior

### PR DESCRIPTION
**Pull request by philsturgeon**
OpenAPI Schema Objects and JSON Schema have a few fundamental differences, and this clears up a bit of confusion about one of them.

Basically, in JSON Schema, a schema of `{}` is totally valid, and when type is missing it essentially means "type can be anything", which would allow `null`. 

With OpenAPI Schema Objects, `{}` should be allowed, but nullable confuses that with its default of false. You'd need to write `{ nullable: true }` at the very least to let a `null` squeak through.

This seems like its breaking, but I dug around and most tooling will fall over if type is missing. 

image

Seeing as some tooling prevents `{}` and some tooling supports it, we have a bit of a nasty situation that needs resolving. This change basically says "nullable adds null to the type that exist and missing type essentially means `type: string|integer|array|etc`... so adding null to that list is fine.

### CODE REVIEW

File: versions/3.0.3.md [hunk 1]
```diff
@@ -2326,7 +2326,7 @@ Other than the JSON Schema subset fields, the following fields MAY be used for f
 ##### Fixed Fields
 Field Name | Type | Description
 ---|:---:|---
-<a name="schemaNullable"></a>nullable | `boolean` | Allows sending a `null` value for the defined schema. Default value is `false`.
+<a name="schemaNullable"></a>nullable | `boolean` | Allows sending a `null` value for the defined schema. Default value is `false`.<br><br>A `nullable` value of `true` allows `null` in addition to other specified `type` values only if `nullable` and `type` are both specified within the same schema object. The effect of `"nullable" : true` is limited to its expansion of the set of allowed types, within the scope of its containing schema. An `enum`, `const`, `allOf`, or other keyword can independently prohibit `null` values, effectively overriding `"nullable" : true`.<br><br>A `nullable` value of `false` leaves the specified or default `type` unmodified. It has no effect on the schema object.
```
Comment (tedepstein) [inline versions/3.0.3.md]: This is the revised language we agreed to on the TSC call: 

```suggestion
<a name="schemaNullable"></a>nullable | `boolean` | A `true` value expands the allowed type defined by the `type` keyword, only if `type` is explicitly defined within the same Schema Object. Other Schema Object constraints retain their defined behavior, and therefore may disallow the use of `null` as a value. A `false` value leaves the specified or default `type` unmodified. The default value is `false`.
```

Comment (tedepstein) [reply versions/3.0.3.md]: @darrelmiller , @webron , @handrews, @philsturgeon , please note the following edits to our edit:

A `true` value ~expands~ **adds `"null"` to** the allowed type ~defined~ **specified** by the `type` keyword, only if `type` is explicitly defined within the same Schema Object. Other Schema Object constraints retain their defined behavior, and therefore may disallow the use of `null` as a value. A `false` value leaves the specified or default `type` unmodified. The default value is `false`.

Somehow in our collaborative real-time editing, we neglected to state that `nullable` adds `"null"` to the set of allowed types. We just said that it "expands" the `type` value, but didn't state the obvious (to us) nature of said expansion.

I have corrected that here and in the proposal in PR #2050.

File: versions/3.0.3.md [hunk 2]
```diff
@@ -2326,7 +2326,7 @@ Other than the JSON Schema subset fields, the following fields MAY be used for f
 ##### Fixed Fields
 Field Name | Type | Description
 ---|:---:|---
-<a name="schemaNullable"></a>nullable | `boolean` | Allows sending a `null` value for the defined schema. Default value is `false`.
+<a name="schemaNullable"></a>nullable | `boolean` | A `true` value adds `"null"` to the allowed type specified by the `type` keyword, only if `type` is explicitly defined within the same Schema Object. Other Schema Object constraints retain their defined behavior, and therefore may disallow the use of `null` as a value. A `false` value leaves the specified or default `type` unmodified. The default value is `false`.
```
Comment (philsturgeon) [inline versions/3.0.3.md]: Do we also need to mention and clear up that type being missing is string, integer, null, array, etc?

Comment (tedepstein) [reply versions/3.0.3.md]: > Do we also need to mention and clear up that type being missing is string, integer, null, array, etc?

I think that's covered by JSON Schema. `type` is a constraint, and in the absence of a `type` constraint, values of any type are allowed. 

I would not mind stating it explicitly, but TSC folks on the call were pushing to keep the spec minimal and concise. Paraphrasing what was said on the call, people will inevitably get confused and ask questions, and no amount of explanatory verbiage in the spec will prevent that from happening.

Comment (handrews) [reply versions/3.0.3.md]: @philsturgeon we covered this in the call.  We want to keep the text minimal, and defer to JSON Schema as much as possible.

**Comment by tedepstein** [comment]
@darrelmiller , this is the PR that we would like to merge for the 3.0.3 patch release. 

I believe we have discussed this exhaustively on TSC calls. The new `nullable` definition language is what we composed in a real-time editing session with @webron, @handrews, and others on a TSC call a few weeks ago. The complete background, motivation, and analysis are documented in the [proposal here](https://github.com/OAI/OpenAPI-Specification/blob/master/proposals/003_Clarify-Nullable.md).

As discussed, please herd the TSC cats to approve and merge.  Thank you!  :-)

**Comment by handrews** [comment]
@darrelmiller my recollection matches @tedepstein's.

I also put wording as close to exactly matching this as possible given `type` arrays in #1977, in case anyone is worried about wether/how this would fit with possible OAS 3.1 changes.

**Comment by philsturgeon** [comment]
giphy

**Comment by darrelmiller** [comment]
I think it is fair to say we have done our due diligence on this particular issue.   Hey @OAI/tsc please review and and sign off so we can get this merged into 3.0.3 so we can clear the path to updating JSON Schema in 3.1.  Ted's proposal provides all the background to this conversation and the reasoning as to why this solution was chosen.

**Comment by philsturgeon** [comment]
Yaaaaay, @darrelmiller could you click that button?

**Comment by darrelmiller** [comment]
I was going to say that we a majority TSC vote, but I re-read the governance rules and we only need that for merging the patch branch to master.  Sooooooo.  Time to click the big green button.

**Comment by philsturgeon** [comment]
Thank you Darrel!!

**Comment by darrelmiller** [comment]
@philsturgeon No, thank you and all the other community folks who put the effort into solving this tricky problem  /cc @tedepstein @handrews

**Comment by webron** [comment]
@philsturgeon, @tedepstein, @handrews - indeed, thank you all for pushing this forward. I love this came to be an elegant solution!

**Pull request merged**